text/0000-replace-blacklist-whitelist.md

```diff
@@ -0,0 +1,80 @@
+- Start Date: 2020-06-15
+- Relevant Team(s): Ember CLI, Learning
+- RFC PR: [#639](https://github.com/emberjs/rfcs/pull/639)
+- Tracking: (leave this empty)
+
+# Replace terms blacklist & whitelist in Ember CLI
+
+## Summary
+
+Ember.js prides itself (rightly) on being an inclusive framework. To further improve on this, we should remove the terms "blacklist" and "whitelist" with more neutral replacements.
+
+## Motivation
+
+The terms "blacklist" and "whitelist" can be considered racially insenstive. While the origin of these terms in this context is not in itself racially motivated, the fact that in todays context it _can be considered racially insensitive_ has been discussed frequently - see for example [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6148600/), [here](https://bugs.chromium.org/p/chromium/issues/detail?id=981129#c16) or [here](https://www.zdnet.com/article/uk-ncsc-to-stop-using-whitelist-and-blacklist-due-to-racial-stereotyping/).
+
+Other projects have already taken similar steps, for example [Go](https://go-review.googlesource.com/c/go/+/236857/), [Android](<https://android-review.googlesource.com/q/topic:%22soong_inclusive_language%22+(status:open%20OR%20status:merged)>), [curl](https://github.com/curl/curl/pull/5546), [PHPUnit](https://github.com/sebastianbergmann/phpunit/issues/4275), and many more.
+
+These terms are used in Ember CLI for rather advanced functionality, which should make providing a better alternative naming and deprecating the current one rather easy.
+
+## Detailed design
+
+You can use `blacklist` and `whitelist` to include/exclude certain addons from your build:
+
+https://ember-cli.com/user-guide/#whitelisting-and-blacklisting-assets
+
+We should replace these with `exclude` or `include`:
```

Comment:
include is, unfortunately, ambiguous, for example: "please include an Oreo MilkShake with my order." It does not limit my order to an Oreo MilkShake; rather, it implies my order will have at-least one Oreo MilkShake.

The existing whitelist implementation does not guarantee the existence of any of these mentioned addons; rather, the whitelist limits addons to the ones mentioned in the provided configuration. It is also worth noting; the existing implementation does also not require any of the "whitelisted" addons to be present.

Given that, I would recommend a more concise name, and the best I could come up with is `allowed,` which pairs nicely with `disallowed.`

Reply:
I don't really agree on this. I went to Ember Observer to check how often certain terms are used in the community. For easier comparison, I restricted searches to index.js - these were the results:

- blacklist: 16 addons
- whitelist: 20 addons
- allowlist, allowedlist, denylist, deniedlist, disallowlist, disallowedlist, blockedlist: 0 addons
- blocklist: 2 addons
- safelist: 1 addon
- allow: 4 addons
- deny: 1 addon
- allowed: 18 addons
- disallowed: 1 addon
- include: 207 addons
- exclude: 101 addons

Also, looking at some popular usages like broccoli-funnel, the general mode of function is pretty much the same. E.g. I can specify:

```js
funnel('src', {
  include: ['my-file', 'my-non-existing-file']
});
```

or

```js
funnel('src', {
  exclude: ['my-file']
});
```

Yes, I know that takes a glob pattern, but the mental model is the same for me.

So for me, it would feel quite natural to use the same wording for configuration of the addons.
I don't know of any occurrence in an Ember app where I used `include` in a way you described (meaning it adds this to a selection/list at least once), but I've used it frequently in the way broccoli-funnel uses it.

Reply:
As a maintainer of broccoli-funnel I apologies for misleading you and can tell you with confidence, if it where to be written again it would not use `include`. Rather it would mimic `walkSync` `globs` + `ignore` verbiage.

Reply:
`disallowed` seems like a good solution. I agree that `include` / `exclude` can be ambiguous, and have seen this problem crop up repeatedly in various OSS projects I've worked on.

There's another ambiguity than the one you referenced: what happens if you use both of the features together. I tend to not think that it's important to have a pair that sounds symmetrical, because it intrinsically opens up this question.

Spitballing, I prefer a mutually exclusive pair like `only` and `except`. The word "only" strongly implies that the other files aren't included. The word "except" strongly implies that the other files are included. The words "include" and "exclude" are much more neutral, which gives rise to the "what happens if both are used" confusion.

Reply:
(btw, @mydea the fact that an idiom is popular doesn't mean it's not confusing or misleading -- in fact, that's the whole premise of this PR!)

Reply:
My only concern with `only` is that the list of addons can include add-ons which are not present. This lead me to `allowed`/ `disallow` but, that is a minor detail and could just be how I interpret `only`

Reply:
I am also OK with using a different term - I chose those to be consistent with current configuration options used in the Ember community, and because they made sense to me. 

I would prefer `only`/`except` over `allowed`/`disallow`, also because the latter can be tricky to get right (`allowed` vs. `allow` vs. `disallow` vs. `disallowed`).

Another option could be `only`/`ignore`, maybe.

Reply:
@wycats after discussion on Friday, do we want to go with `allowlist` and `disallowlist` ?

Reply:
Maybe I'm late to the party, but how about `includeOnly` and `exclude`?